File: FortranUtilities/chebyshev.f90
    
module chebyshev
    use constants
    interface getPhi
        module procedure getPhiScalar, getPhiVect
    end interface  
    contains 
 
    !Evaluates the 1st S Cheybshev polynomials at each point in xvect
    function getPhiVect(xvect,S)
        implicit none
        
        real(dbl), intent(in):: xvect(:)
        integer, intent(in):: S
        real(dbl) getPhiVect(size(xvect),S)
        integer i
        
        getPhiVect(:,1) = 1
        if (S == 1) return
        getPhiVect(:,2) = xvect
        do i=3,S
            getPhiVect(:,i) = 2*xvect*getPhiVect(:,i-1) - getPhiVect(:,i-2)
        end do
        getPhiVect = merge(getPhiVect,1.0,getPhiVect<1.0)
        getPhiVect = merge(getPhiVect,-1.0,getPhiVect>-1.0)
    end function getPhiVect
    
    !Evaluates the 1st S Cheybshev polynomials at x
    function getPhiScalar(x,S)
        implicit none
        
        real(dbl), intent(in):: x
        integer, intent(in):: S
        real(dbl) getPhiScalar(S)
        integer i
        
        getPhiScalar(1) = 1
        if (S == 1) return
        getPhiScalar(2) = x
        do i=3,S
            getPhiScalar(i) = 2*x*getPhiScalar(i-1) - getPhiScalar(i-2)
        end do
        getPhiScalar = merge(getPhiScalar,1.0,getPhiScalar<1.0)
        getPhiScalar = merge(getPhiScalar,-1.0,getPhiScalar>-1.0)
    end function getPhiScalar
    
    !Evaluates the nth Cheybshev polynomials at each point in xvect
    function getCheb(xvect,n)
        implicit none
        
        real(dbl), intent(in):: xvect(:)
        integer, intent(in):: n
        real(dbl) getCheb(size(xvect))
             
        getCheb = cos((n-1)*acos(xvect))
        
    end function getCheb
    
    !Returns the s extrema of the of sth Chebyshev polynomial (also called the Gauss-Lobatto nodes)
    function getExtrema(s)
            integer s, j
            real(dbl) getExtrema(s)
            if (s == 1) then
                getExtrema(1) = 0.0d0
            else
                getExtrema =(/ ( -cos(pi*(j-1)/(s-1)), j=1,s) /)
            end if
    end function getExtrema
    
    !Returns the s roots of the of sth Chebyshev polynomial (also called the Gauss-Lobotto nodes)
    function getRoots(s)
            integer s, i
            real(dbl) getRoots(s)

            getRoots =(/( -cos((2*i-1)*pi/2/s) ,i=1,s)/)
 
    end function getRoots                
end module chebyshev    

File: C++Utilities/chebyshev.cpp
#include <cmath>
#include "kk_constant.hpp"
#include "dw_dense_matrix.hpp"
#include "chebyshev.hpp"

using namespace std; 

// Evaluates the 1st S Cheybshev polynomials at x
TDenseVector getPhi(double x, int S)
{
	if (S <= 0)
 		return TDenseVector(0); 

	TDenseVector phi(S,1.0); 
	if (S == 1)
		return phi; 
	phi[1] = x; 
	for (int i=2; i<S; i++)
	{
		phi[i] = 2*x*phi[i-1]-phi[i-2]; 
		phi[i] = phi[i] < 1.0 ? phi[i] : 1.0; 
		phi[i] = phi[i] > -1.0 ? phi[i] : -1.0; 
	}
	return phi; 
}

// Evaluates the 1st S Cheybshev polynomials at each point in xvect
TDenseMatrix getPhi(const TDenseVector &x, int S)
{
	if (x.dim <= 0 || S <= 0)
		return TDenseMatrix(0,0); 

	TDenseMatrix phi(x.dim, S, 1.0); 
	if (S == 1)
		return phi; 
	phi.InsertColumnMatrix(0,1,x); // phi(:,1) = x; 
	for (int i=2; i<S; i++)
	{
		phi.InsertColumnMatrix(0,i,(2.0*DotMultiply(x,phi.ColumnVector(i-1))-phi.ColumnVector(i-2)));
		for (int j=0; j<x.dim; j++)
		{
			phi(j,i) = phi(j,i) < 1.0 ? phi(j,i) : 1.0; 
			phi(j,i) = phi(j,i) > -1.0 ? phi(j,i) : 1.0; 
		}
	}
	return phi; 
}

// Evaluates the nth Cheybshev polynomials at each point in xvect
TDenseVector getCheb(const TDenseVector &x, int n)
{
	TDenseVector cheb(x.dim,0.0); 
	for (int i=0; i< x.dim; i++)
		cheb[i] = cos((n-1.0)*acos(x[i])); 
	return cheb; 
}

// Returns the s extrema of the of sth Chebyshev polynomial (also called the Gauss-Lobatto nodes)
TDenseVector getExtrema(int s)
{
	TDenseVector extrema(s,0.0); 
	for (int j=0; j<s; j++)
		extrema[j] = -cos(PI*j/(s-1.0)); 
	return extrema; 
}

// Returns the s roots of the of sth Chebyshev polynomial (also called the Gauss-Lobotto nodes)
TDenseVector getRoots(int s)
{
	TDenseVector roots(s,0.0); 
	for (int i=0; i<s; i++)
		roots[i] = -cos( (2.0*i+1.0)*PI/(2.0*s)); 
	return roots; 
}

TDenseVector getPhiScalar(double x, int S) { return getPhi(x,S); }
TDenseMatrix getPhiVect(const TDenseVector &x, int S) { return getPhi(x,S); }
